lib/src/visitor/async_evaluate.dart

```diff
@@ -1987,9 +1993,13 @@ class _EvaluateVisitor
   Future<Value?> visitWarnRule(WarnRule node) async {
     var value =
         await _addExceptionSpanAsync(node, () => node.expression.accept(this));
-    _logger.warn(
-        value is SassString ? value.text : _serialize(value, node.expression),
-        trace: _stackTrace(node.span));
+    var message =
+        value is SassString ? value.text : _serialize(value, node.expression);
+
+    // Don't use [_warn] because we don't want to pass the span to the logger.
+    if (_warningsEmitted.add(Tuple2(message, node.span))) {
```

Comment:
should `@warn` be deduplicated too ?

For warnings triggered in dart code, the span is the place that you actually need to change to resolve the warning (at least in most cases).
For userland functions triggering a warning with `@warn` when being misused, the span used here would be the location of the `@warn` rule inside the function, not the location of the caller (which is what where the change would have to be done).

Reply:
Hmm, that's a good point. We don't necessarily have a good understanding of where the boundary is between "user code" and "library code" such that we can reliably tell which stack frame has the span that should actually be used for uniqueness here. I'll drop the `@warn` de-duplication.